### DISCUSSION TITLE
How to insert a next/link into a html string

### DISCUSSION THREAD

**Discussion by Glutch** [vercel/next.js#35941] [Help] [answered]
I'm currently looking through a string, finding hashtags and replacing them with a link.

```javascript
const tags = text.match(/(#[a-z]+)/gi) || []
const parsed = text
      .split(/(#[a-z]+)/gi)
      .map(word =>
        tags.includes(word)
          ? `<a href="/t/${word.replace('#', '')}" style="color: ${get_tag_color(
              word
            )}; cursor: pointer; font-weight: bold;">${word}</a>`
          : word
      )
```

I am then displaying this parsed string like this
`<div dangerouslySetInnerHTML={{__html: parsed_text}}></div>`

As you can tell, i'm struggling to insert an actual nextjs Link into this.

I found another guy with the same issue 1 year ago with no answers: https://stackoverflow.com/questions/65090429/inserting-next-js-links-into-an-html-string

Any help greatly appreciated

**Comment by icyJoseph** [comment, marked as answer]
Doing this is not trivial, I think.

If we break down the problem, first of all, you have a source string that ought to be modelled as objects. Second, you need to walk these objects to render them as React Elements. Third, whenever you encounter an `a` tag, you'd return a `next/link` wrapping the `a` tag. 

You could experiment with something like https://www.npmjs.com/package/html-react-parser - which at the cost of many kb's can help you out (10 kb according to https://bundlephobia.com/package/html-react-parser@1.4.10)

Matching html with regex is very much not recommended. 

Taking html strings is always a matter to be treated carefully. XSS is nasty. I am not sure about the source of your string, but if you could model it as objects (nearly like React object, with children an all and prop-like attributes), you'd be doing yourself a favour.

**Reply by Glutch** [reply]
Thank you! i managed to solve it with html-react-parser and i do not need to dangerouslySetInnerHTML.
I do however still use my initial parsed function with regex and after that parse it again with html-react-parser.

How would you solve this without regex? I can't think of another solution. I need to both find #hashtags and links (https://abc.xyz) from a user inputted string

**Reply by icyJoseph** [reply]
Oh I thought you were parsing the html with regex, but you target links. It might be alright! Though users might still be able to craft a string that causes issues.

With that parsing library you can make sure to only run your function on href's right? That should be pretty safe I think.

Don't forget to mark as answered 🙂

**Reply by k-kombinat** [reply]
@Glutch How you managed the problem with nextjs and html-react-parser?

On my site this is not workin:

```
import parse from 'html-react-parser'

const link = parse('<Link href="/settings"><a>Test</a></Link>')
 
// will rendered as

<link href="<host>/settings"/><a>Test</a>

```

What do I miss?

**Reply by icyJoseph** [reply]
I think OP didn't have the Link tag on the text, but rather only anchor tags that they wanted to wrap with next/link. There must be an option to tell the parser to keep relative links, well, relative.

Sounds like a sanitize thing done by the library. Might get some time to check this out next week.

Going over it once again, it looks like you need to pass a second argument to the parser:

```js
parse('<p id="replace">text</p>', {
  replace: domNode => {
    if (domNode.attribs && domNode.attribs.id === 'replace') {
      return <span>replaced</span>;
    }
  }
});
```

In your case, you want to look for domNode with name Link I suppose, and manually pass attributes to it.

**Reply by k-kombinat** [reply]
Thx 4 the reply. I will try this out.

**Comment by gitamj** [comment]
I was able to make it work using this awesome package: https://www.npmjs.com/package/html-to-react

Here is a working demo https://github.com/gitamj/html-string-to-nextjs
